worldedit-sponge/src/main/java/com/sk89q/worldedit/sponge/SpongeWorld.java

```diff
@@ -19,204 +19,358 @@
         } else {
             throw new RuntimeException("The reference to the world was lost (i.e. the world may have been unloaded)");
         }
     }
 
+    // This is sus but leaving it for later world name/id reworks
     @Override
     public String getName() {
-        return getWorld().getName();
+        return getWorld().key().asString();
     }
 
     @Override
     public String getId() {
-        return getName().replace(" ", "_").toLowerCase(Locale.ROOT)
-            + getWorld().getDimension().getType().getName().toLowerCase(Locale.ROOT);
+        return getWorld().key().asString();
     }
 
     @Override
     public Path getStoragePath() {
-        return getWorld().getDirectory();
+        return getWorld().directory();
+    }
+
+    @Override
+    public BlockState getBlock(BlockVector3 position) {
+        return SpongeAdapter.adapt(getWorld().block(
+            position.getX(), position.getY(), position.getZ()
+        ));
+    }
+
+    @Override
+    public BaseBlock getFullBlock(BlockVector3 position) {
+        CompoundTag entity = getWorld()
+            .blockEntity(position.getX(), position.getY(), position.getZ())
+            .map(e -> NbtAdapter.adaptToWorldEdit(e.createArchetype().blockEntityData()))
+            .orElse(null);
+        return getBlock(position).toBaseBlock(entity);
     }
 
-    @SuppressWarnings("WeakerAccess")
-    protected BlockState getBlockState(BlockStateHolder<?> block) {
-        if (block instanceof com.sk89q.worldedit.world.block.BlockState) {
-            BlockState state = Sponge.getRegistry().getType(BlockType.class, block.getBlockType().getId()).orElse(BlockTypes.AIR).getDefaultState();
-            for (Map.Entry<Property<?>, Object> entry : block.getStates().entrySet()) {
-                // TODO Convert across states
+    @Override
+    public <B extends BlockStateHolder<B>> boolean setBlock(BlockVector3 position, B block, SideEffectSet sideEffects) throws WorldEditException {
+        checkNotNull(position);
+        checkNotNull(block);
+
+        ServerWorld world = getWorld();
+
+        org.spongepowered.api.block.BlockState newState = SpongeAdapter.adapt(block.toImmutableState());
+
+        boolean didSet = world.setBlock(
+            position.getX(), position.getY(), position.getZ(),
+            newState,
+            BlockChangeFlags.NONE
+                .withUpdateNeighbors(sideEffects.shouldApply(SideEffect.NEIGHBORS))
+                .withNotifyClients(true)
+                .withPhysics(sideEffects.shouldApply(SideEffect.UPDATE))
+                .withNotifyObservers(sideEffects.shouldApply(SideEffect.UPDATE))
+                .withLightingUpdates(sideEffects.shouldApply(SideEffect.LIGHTING))
+                .withPathfindingUpdates(sideEffects.shouldApply(SideEffect.ENTITY_AI))
+                .withNeighborDropsAllowed(false)
+                .withBlocksMoving(false)
+                .withForcedReRender(false)
+                .withIgnoreRender(false)
```

Comment:
I wonder if this should be cached -- worth measuring performance difference if you use e.g. an `EnumMap` to store pre-calculated flags.